.github/scripts/simple-smoke-test.sh

```diff
@@ -1,49 +1,88 @@
+}
+
+function remove_container_if_exists() {
+    local container_name=$1
+
+    local containers
+    containers=$(docker ps --all --quiet --filter name="${container_name}")
+
+    if [[ -n ${containers} ]]; then
+      echo "Removing existing '${container_name}' container"
+      docker container rm --force "${container_name}"
     fi
+}
 
-    echo "Checking if $image is EE"
-    if docker run --rm $image bash -c 'compgen -G lib/*enterprise*'; then
+function check_distribution_type() {
+    local image=$1
+    local expected_distribution_type=$2
+
+    echo "Checking ${image} distribution type"
+    if docker run --rm "${image}" bash -c 'compgen -G lib/*enterprise*'; then
       echo "EE contents identified"
       distribution_type="ee"
     else
       echo "No EE contents identified - assuming OSS"
       distribution_type="oss"
     fi
 
-    if [[ "$distribution_type" != "$expected_distribution_type" ]]; then
-      echo "Distribution was $distribution_type, not $expected_distribution_type as expected"
+    if [[ "${distribution_type}" != "${expected_distribution_type}" ]]; then
+      echo "Image ${image} should contain ${expected_distribution_type} distribution but ${distribution_type} was detected"
       exit 1
     fi
+}
+
+function check_version() {
+    local image=$1
+    local expected_version=$2
+
+    echo "Checking ${image} version"
+    local version
+    version=$(docker run --rm "${image}" bin/hz-cli --version | awk '/Hazelcast/ {print $2}')
+    if [[ "${version}" == "${expected_version}" ]]; then
+      echo "${image} version identified as ${version}"
+    else
+      echo "${image} version was ${version}, not ${expected_version} as expected"
+      exit 1
+    fi
+}
+
+function test_map_read_write() {
+    local image=$1
+    local expected_distribution_type=$2
 
-    echo "Starting container '$container_name' from image '$image'"
-    docker run -it --name "$container_name" -e HZ_LICENSEKEY -e HZ_INSTANCETRACKING_FILENAME -d -p5701:5701 "$image"
+    echo "Starting container '${container_name}' from image '${image}'"
+    docker run -it --name "${container_name}" -e HZ_LICENSEKEY -e HZ_INSTANCETRACKING_FILENAME -d -p5701:5701 "${image}"
     local key="some-key"
     local expected="some-value"
-    echo "Putting value '$expected' for key '$key'"
-    while ! clc --timeout 5s map set -n some-map $key $expected --log.path stderr
+    echo "Putting value '${expected}' for key '${key}'"
+    while ! clc --timeout 5s map set -n some-map "${key}" "${expected}" --log.path stderr
     do
       echo "Retrying..."
       sleep 3
     done
-    echo "Getting value for key '$key'"
+    echo "Getting value for key '${key}'"
     local actual
-    actual=$(clc map get --format delimited -n some-map $key --log.path stderr)
-    echo "Stopping container $container_name}"
-    docker stop "$container_name"
+    actual=$(clc map get --format delimited -n some-map "${key}" --log.path stderr)
```

Comment:
The command itself could fail for whatever reason like HZ is not responding
I think we should distinguish that vs no value found

Reply:
> The command itself could fail for whatever reason like HZ is not responding I think we should distinguish that vs no value found

I _think_ it'll work the same - whether the command prints `0`, `1` or `Failed to connect to Hazelcast cluster!`, we'll check that against `expected` and error & print if different.

Reply:
> Understood but I am saying don't rely on `expected` later on if the issue is with cluster/docker then we should error out so its clear when debugging I think we should not conflate the getting and then the checking Thats how I normally do it

Sorry, I don't understand?

Reply:
I am saying we should check for error after running `clc` and bail out if error rather then relying on the later steps when checking for key/value

Reply:
> I am saying we should check for error after running `clc` and bail out if error rather then relying on the later steps when checking for key/value

What error would you be looking to catch?

`install_clc` _should_ fail if the installation fails so beyond that we can assume it's working.

This is existing functionality btw.

Reply:
any error from `clc` tbh
I think using `$(clc...)` captures the error and I am thinking script will not fail unless you check the error
May be I am wrong

Reply:
> any error from `clc` tbh I think using `$(clc...)` captures the error and I am thinking script will not fail unless you check the error May be I am wrong

but `actual` content will be wrong and test will fail

There's also `set -E` flag to catch subshell error if I'm not mistaken, see https://stackoverflow.com/a/72103511

Reply:
yes it will fail as `actual` content will be wrong
But normal practice is to handle errors right where it happens
If `set -E` works then thats ok too
Otherwise, don't want to drag so happy as it is!

Reply:
> yes it will fail as `actual` content will be wrong But normal practice is to handle errors right where it happens If `set -E` works then thats ok too Otherwise, don't want to drag so happy as it is!

I've tested it with a simple bash script:
```
#!/usr/bin/env bash

set -o errexit -o nounset -o pipefail

actual=$(clc map get --format delimited -n some-map this-map-doesnt-exist --log.path stderr)
echo finished
```

And it sets the parent errorlevel when the script exits

Reply:
thanks for confirming. will resolve

Reply:
> thanks for confirming. will resolve

Thanks - I think that was the last outstanding issue from you?

Reply:
Last one https://github.com/hazelcast/hazelcast-docker/pull/776/files#r1668945722
Just commented